src/dataset_tools/bq_squashfs/download_images.py

```
@@ -0,0 +1,627 @@
+    Retries on rate-limit (429), transient server errors (5xx),
+    connection errors (including Errno 16 — too many open sockets),
+    and timeouts.  Raises on permanent client errors (4xx except 429)
+    or after exhausting all retries.
+    """
+    session = _get_session()
+    for attempt in range(_MAX_RETRIES + 1):
+        try:
+            resp = session.get(url, timeout=30, stream=True)
+            if resp.status_code in _RETRY_STATUSES and attempt < _MAX_RETRIES:
+                delay = min(_BACKOFF_BASE * (2**attempt), _BACKOFF_MAX)
+                delay += random.uniform(0, delay * 0.25)
```

Ruff (0.15.15)

[error] Standard pseudo-random generators are not suitable for cryptographic purposes

(S311)